File: backend/app/schemas/dataset.py
from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class IngestMode(str, Enum):
    CREATE = "create"
    OVERWRITE = "overwrite"


class DatasetIngestRequest(BaseModel):
    source_id: UUID
    table_or_file: str = Field(min_length=1)
    dataset_name: str = Field(min_length=1, max_length=255)
    display_name: str = Field(min_length=1, max_length=255)
    mode: IngestMode = IngestMode.CREATE


class DatasetUpdate(BaseModel):
    display_name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None


class DatasetResponse(BaseModel):
    id: UUID
    name: str
    display_name: str
    description: str | None = None
    source_id: UUID | None = None
    iceberg_namespace: str
    iceberg_table: str
    iceberg_location: str
    schema_json: dict[str, Any] | None = None
    row_count: int | None = None
    size_bytes: int | None = None
    last_synced_at: datetime | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class PreviewResponse(BaseModel):
    columns: list[str]
    rows: list[dict[str, Any]]
    total: int


class ColumnProfile(BaseModel):
    name: str
    type: str
    distinct_count: int | None = None
    null_count: int | None = None
    non_null_count: int | None = None
    min_value: Any | None = None
    max_value: Any | None = None


class ProfileResponse(BaseModel):
    row_count: int
    columns: list[ColumnProfile]

File: frontend/src/types/dataset.ts
export interface Dataset {
  id: string;
  name: string;
  display_name: string;
  description: string | null;
  source_id: string | null;
  iceberg_namespace: string;
  iceberg_table: string;
  iceberg_location: string;
  schema_json: { fields?: { name: string; type: string }[] } | null;
  row_count: number | null;
  size_bytes: number | null;
  last_synced_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface DatasetPreview {
  columns: string[];
  rows: Record<string, unknown>[];
  total: number;
}

export interface ColumnProfile {
  name: string;
  type: string;
  distinct_count: number | null;
  null_count: number | null;
  non_null_count: number | null;
  min_value: unknown;
  max_value: unknown;
}

export interface DatasetProfile {
  row_count: number;
  columns: ColumnProfile[];
}

export interface IngestRequest {
  source_id: string;
  table_or_file: string;
  dataset_name: string;
  display_name: string;
  mode: "create" | "overwrite";
}
